// @tamagui/react-native-web-lite/dist/esm/vendor/react-native/Animated/NativeAnimatedModule.js
import * as TurboModuleRegistry from "../TurboModule/TurboModuleRegistry";
const EventMapping = {}, AnimatedNodeConfig = {}, AnimatingNodeConfig = {}, SaveValueCallback = () => {
}, EndCallback = () => {
}, spec = {
  startOperationBatch: () => {
  },
  finishOperationBatch: () => {
  },
  createAnimatedNode: (tag, config) => {
  },
  updateAnimatedNodeConfig: (tag, config) => {
  },
  getValue: (tag, saveValueCallback) => {
  },
  startListeningToAnimatedNodeValue: (tag) => {
  },
  stopListeningToAnimatedNodeValue: (tag) => {
  },
  connectAnimatedNodes: (parentTag, childTag) => {
  },
  disconnectAnimatedNodes: (parentTag, childTag) => {
  },
  startAnimatingNode: (animationId, nodeTag, config, endCallback) => {
  },
  stopAnimation: (animationId) => {
  },
  setAnimatedNodeValue: (nodeTag, value) => {
  },
  setAnimatedNodeOffset: (nodeTag, offset) => {
  },
  flattenAnimatedNodeOffset: (nodeTag) => {
  },
  extractAnimatedNodeOffset: (nodeTag) => {
  },
  connectAnimatedNodeToView: (nodeTag, viewTag) => {
  },
  disconnectAnimatedNodeFromView: (nodeTag, viewTag) => {
  },
  restoreDefaultValues: (nodeTag) => {
  },
  dropAnimatedNode: (tag) => {
  },
  addAnimatedEventToView: (viewTag, eventName, eventMapping) => {
  },
  removeAnimatedEventFromView: (viewTag, eventName, animatedNodeTag) => {
  },
  // Events
  addListener: (eventName) => {
  },
  removeListeners: (count) => {
  },
  // All of the above in a batched mode
  queueAndExecuteBatchedOperations: (operationsAndArgs) => {
  }
};
var NativeAnimatedModule_default = TurboModuleRegistry.get("NativeAnimatedModule") || spec;
export {
  AnimatedNodeConfig,
  AnimatingNodeConfig,
  EndCallback,
  EventMapping,
  SaveValueCallback,
  NativeAnimatedModule_default as default
};
