// @tamagui/react-native-web-lite/dist/esm/vendor/react-native/Animated/NativeAnimatedTurboModule.native.js
import * as TurboModuleRegistry from "../TurboModule/TurboModuleRegistry.native.js";
var EventMapping = {},
  AnimatedNodeConfig = {},
  AnimatingNodeConfig = {},
  SaveValueCallback = function () {},
  EndCallback = function () {},
  spec = {
    startOperationBatch: function () {},
    finishOperationBatch: function () {},
    createAnimatedNode: function (tag, config) {},
    getValue: function (tag, saveValueCallback) {},
    startListeningToAnimatedNodeValue: function (tag) {},
    stopListeningToAnimatedNodeValue: function (tag) {},
    connectAnimatedNodes: function (parentTag, childTag) {},
    disconnectAnimatedNodes: function (parentTag, childTag) {},
    startAnimatingNode: function (animationId, nodeTag, config, endCallback) {},
    stopAnimation: function (animationId) {},
    setAnimatedNodeValue: function (nodeTag, value) {},
    setAnimatedNodeOffset: function (nodeTag, offset) {},
    flattenAnimatedNodeOffset: function (nodeTag) {},
    extractAnimatedNodeOffset: function (nodeTag) {},
    connectAnimatedNodeToView: function (nodeTag, viewTag) {},
    disconnectAnimatedNodeFromView: function (nodeTag, viewTag) {},
    restoreDefaultValues: function (nodeTag) {},
    dropAnimatedNode: function (tag) {},
    addAnimatedEventToView: function (viewTag, eventName, eventMapping) {},
    removeAnimatedEventFromView: function (viewTag, eventName, animatedNodeTag) {},
    // Events
    addListener: function (eventName) {},
    removeListeners: function (count) {}
  },
  NativeAnimatedTurboModule_default = TurboModuleRegistry.get("NativeAnimatedTurboModule") || spec;
export { AnimatedNodeConfig, AnimatingNodeConfig, EndCallback, EventMapping, SaveValueCallback, NativeAnimatedTurboModule_default as default };
