// @tamagui/react-native-web-lite/dist/cjs/vendor/react-native/LayoutAnimation/index.native.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf, __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: !0 }) : target,
  mod
)), __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var LayoutAnimation_exports = {};
__export(LayoutAnimation_exports, {
  default: () => LayoutAnimation_default
});
module.exports = __toCommonJS(LayoutAnimation_exports);
var import_Platform = __toESM(require("../../../exports/Platform")), import_UIManager = __toESM(require("../../../exports/UIManager")), __DEV__ = process.env.NODE_ENV !== "production";
function configureNext(config, onAnimationDidEnd) {
  import_Platform.default.isTesting || import_UIManager.default.configureNextLayoutAnimation(config, onAnimationDidEnd ?? function() {
  }, function() {
  });
}
function create(duration, type, property) {
  return {
    duration,
    create: {
      type,
      property
    },
    update: {
      type
    },
    delete: {
      type,
      property
    }
  };
}
var Presets = {
  easeInEaseOut: create(300, "easeInEaseOut", "opacity"),
  linear: create(500, "linear", "opacity"),
  spring: {
    duration: 700,
    create: {
      type: "linear",
      property: "opacity"
    },
    update: {
      type: "spring",
      springDamping: 0.4
    },
    delete: {
      type: "linear",
      property: "opacity"
    }
  }
}, LayoutAnimation = {
  /**
  * Schedules an animation to happen on the next layout.
  *
  * @param config Specifies animation properties:
  *
  *   - `duration` in milliseconds
  *   - `create`, `AnimationConfig` for animating in new views
  *   - `update`, `AnimationConfig` for animating views that have been updated
  *
  * @param onAnimationDidEnd Called when the animation finished.
  * Only supported on iOS.
  * @param onError Called on error. Only supported on iOS.
  */
  configureNext,
  /**
  * Helper for creating a config for `configureNext`.
  */
  create,
  Types: Object.freeze({
    spring: "spring",
    linear: "linear",
    easeInEaseOut: "easeInEaseOut",
    easeIn: "easeIn",
    easeOut: "easeOut",
    keyboard: "keyboard"
  }),
  Properties: Object.freeze({
    opacity: "opacity",
    scaleX: "scaleX",
    scaleY: "scaleY",
    scaleXY: "scaleXY"
  }),
  checkConfig() {
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++)
      args[_key] = arguments[_key];
    console.error("LayoutAnimation.checkConfig(...) has been disabled.");
  },
  Presets,
  easeInEaseOut: configureNext.bind(null, Presets.easeInEaseOut),
  linear: configureNext.bind(null, Presets.linear),
  spring: configureNext.bind(null, Presets.spring)
}, LayoutAnimation_default = LayoutAnimation;
