// @tamagui/react-native-web-lite/dist/esm/index.js
import {
  AccessibilityUtil,
  canUseDOM,
  clickProps,
  createDOMProps,
  dismissKeyboard,
  flattenStyle,
  ImageLoader,
  InteractionManager,
  isWebColor,
  LocaleProvider,
  mergeRefs,
  normalizeColor,
  Platform,
  processColor,
  processStyle,
  StyleSheet,
  TextAncestorContext,
  UIManager,
  useEvent,
  useHover,
  useLayoutEffect,
  useMergeRefs,
  usePlatformMethods
} from "@tamagui/react-native-web-internals";
import { default as default2 } from "./createElement/index";
import { default as default3 } from "./NativeModules/index";
import { default as default4 } from "./render/index";
import { default as default5 } from "./vendor/react-native/Animated/Animated";
import { default as default6 } from "./vendor/react-native/Animated/Easing";
import { default as default7 } from "./vendor/react-native/EventEmitter/NativeEventEmitter";
import { default as default8 } from "./AccessibilityInfo/index";
import { default as default9 } from "./Alert/index";
import { default as default10 } from "./Appearance/index";
import { default as default11 } from "./AppRegistry/index";
import { default as default12 } from "./AppState/index";
import { default as default13 } from "./BackHandler/index";
import { default as default14 } from "./Clipboard/index";
import { default as default15, default as default16 } from "./DeviceEmitter";
import { default as default17 } from "./DeviceInfo/index";
import { default as default18 } from "./Dimensions/index";
import { default as default19 } from "./I18nManager/index";
import { default as default20 } from "./Keyboard/index";
import { default as default21 } from "./Linking/index";
import { default as default22 } from "./PanResponder/index";
import { default as default23 } from "./PixelRatio/index";
import { default as default24 } from "./Share/index";
import { default as default25 } from "./Vibration/index";
import { default as default26 } from "./FlatList";
import { default as default27 } from "./SectionList";
import { default as default28 } from "./VirtualizedList";
import { default as default29 } from "./TouchableNativeFeedback";
import {
  default as default30,
  default as default31,
  default as default32
} from "./UnimplementedView";
import { default as default33, default as default34 } from "./TouchableOpacity";
import { default as default35 } from "./TouchableWithoutFeedback";
import { default as default36 } from "./ActivityIndicator/index";
import { default as default37 } from "./Image/index";
import { default as default38 } from "./ImageBackground/index";
import { default as default39 } from "./KeyboardAvoidingView/index";
import { default as default40 } from "./LogBox/index";
import { default as default41 } from "./Modal/index";
import { default as default42 } from "./Pressable/index";
import { default as default43 } from "./RefreshControl/index";
import { default as default44 } from "./SafeAreaView/index";
import { default as default45 } from "./ScrollView/index";
import { default as default46 } from "./StatusBar/index";
import { default as default47 } from "./Text/index";
import { default as default48 } from "./TextInput/index";
import { default as default49 } from "./View/index";
import { default as default50 } from "./useColorScheme/index";
import { useLocaleContext } from "./useLocaleContext/index";
import { default as default51 } from "./useWindowDimensions/index";
function requireNativeComponent(name) {
  return function() {
    return null;
  };
}
const findNodeHandle = (component) => {
  throw new Error("not supported - use ref instead");
};
export {
  default8 as AccessibilityInfo,
  AccessibilityUtil,
  default36 as ActivityIndicator,
  default9 as Alert,
  default5 as Animated,
  default11 as AppRegistry,
  default12 as AppState,
  default10 as Appearance,
  default13 as BackHandler,
  default14 as Clipboard,
  default15 as DeviceEmitter,
  default16 as DeviceEventEmitter,
  default17 as DeviceInfo,
  default18 as Dimensions,
  default30 as DrawerLayoutAndroid,
  default6 as Easing,
  default26 as FlatList,
  default19 as I18nManager,
  default37 as Image,
  default38 as ImageBackground,
  ImageLoader,
  InteractionManager,
  default20 as Keyboard,
  default39 as KeyboardAvoidingView,
  default21 as Linking,
  LocaleProvider,
  default40 as LogBox,
  default41 as Modal,
  default7 as NativeEventEmitter,
  default3 as NativeModules,
  default22 as PanResponder,
  default23 as PixelRatio,
  Platform,
  default42 as Pressable,
  default43 as RefreshControl,
  default44 as SafeAreaView,
  default45 as ScrollView,
  default27 as SectionList,
  default24 as Share,
  default46 as StatusBar,
  StyleSheet,
  default31 as Switch,
  default47 as Text,
  TextAncestorContext,
  default48 as TextInput,
  default33 as Touchable,
  default32 as TouchableHighlight,
  default29 as TouchableNativeFeedback,
  default34 as TouchableOpacity,
  default35 as TouchableWithoutFeedback,
  UIManager,
  default25 as Vibration,
  default49 as View,
  default28 as VirtualizedList,
  canUseDOM,
  clickProps,
  createDOMProps,
  dismissKeyboard,
  findNodeHandle,
  flattenStyle,
  isWebColor,
  mergeRefs,
  normalizeColor,
  processColor,
  processStyle,
  default4 as render,
  requireNativeComponent,
  default2 as unstable_createElement,
  default50 as useColorScheme,
  useEvent,
  useHover,
  useLayoutEffect,
  useLocaleContext,
  useMergeRefs,
  usePlatformMethods,
  default51 as useWindowDimensions
};
