// @tamagui/react-native-web-lite/dist/esm/without-animated.native.js
import { default as default2 } from "./createElement/index.native.js";
import { processColor, Platform, UIManager, canUseDOM, dismissKeyboard, isWebColor, mergeRefs, useMergeRefs, useEvent, useHover, ImageLoader } from "@tamagui/react-native-web-internals";
import { default as default3 } from "./render/index.native.js";
import { default as default4 } from "./NativeModules/index.native.js";
import { default as default5 } from "./vendor/react-native/EventEmitter/NativeEventEmitter.native.js";
import { default as default6 } from "./AccessibilityInfo/index.native.js";
import { default as default7 } from "./Alert/index.native.js";
import { default as default8 } from "./Appearance/index.native.js";
import { default as default9 } from "./AppRegistry/index.native.js";
import { default as default10 } from "./AppState/index.native.js";
import { default as default11 } from "./BackHandler/index.native.js";
import { default as default12 } from "./Clipboard/index.native.js";
import { default as default13 } from "./DeviceInfo/index.native.js";
import { default as default14 } from "./DeviceEmitter.native.js";
import { default as default15 } from "./DeviceEmitter.native.js";
import { default as default16 } from "./Dimensions/index.native.js";
import { default as default17 } from "./I18nManager/index.native.js";
import { default as default18 } from "./Keyboard/index.native.js";
import { default as default19 } from "./Linking/index.native.js";
import { default as default20 } from "./PanResponder/index.native.js";
import { default as default21 } from "./PixelRatio/index.native.js";
import { default as default22 } from "./Share/index.native.js";
import { default as default23 } from "./Vibration/index.native.js";
import { default as default24 } from "./UnimplementedView.native.js";
import { default as default25 } from "./UnimplementedView.native.js";
import { default as default26 } from "./UnimplementedView.native.js";
import { default as default27 } from "./UnimplementedView.native.js";
import { default as default28 } from "./UnimplementedView.native.js";
import { default as default29 } from "./UnimplementedView.native.js";
import { default as default30 } from "./UnimplementedView.native.js";
import { default as default31 } from "./TouchableOpacity.native.js";
import { default as default32 } from "./TouchableWithoutFeedback.native.js";
import { default as default33 } from "./ActivityIndicator/index.native.js";
import { default as default34 } from "./Image/index.native.js";
import { default as default35 } from "./ImageBackground/index.native.js";
import { default as default36 } from "./KeyboardAvoidingView/index.native.js";
import { default as default37 } from "./Modal/index.native.js";
import { default as default38 } from "./Pressable/index.native.js";
import { default as default39 } from "./RefreshControl/index.native.js";
import { default as default40 } from "./SafeAreaView/index.native.js";
import { default as default41 } from "./ScrollView/index.native.js";
import { default as default42 } from "./StatusBar/index.native.js";
import { default as default43 } from "./Text/index.native.js";
import { default as default44 } from "./TextInput/index.native.js";
import { default as default45 } from "./View/index.native.js";
import { default as default46 } from "./LogBox/index.native.js";
import { default as default47 } from "./useColorScheme/index.native.js";
import { useLocaleContext } from "./useLocaleContext/index.native.js";
import { default as default48 } from "./useWindowDimensions/index.native.js";
export * from "@tamagui/react-native-web-internals";
function requireNativeComponent(name) {
  return function () {
    return null;
  };
}
export { default6 as AccessibilityInfo, default33 as ActivityIndicator, default7 as Alert, default9 as AppRegistry, default10 as AppState, default8 as Appearance, default11 as BackHandler, default12 as Clipboard, default14 as DeviceEmitter, default15 as DeviceEventEmitter, default13 as DeviceInfo, default16 as Dimensions, default24 as DrawerLayoutAndroid, default27 as FlatList, default17 as I18nManager, default34 as Image, default35 as ImageBackground, ImageLoader, default18 as Keyboard, default36 as KeyboardAvoidingView, default19 as Linking, default46 as LogBox, default37 as Modal, default5 as NativeEventEmitter, default4 as NativeModules, default20 as PanResponder, default21 as PixelRatio, Platform, default38 as Pressable, default39 as RefreshControl, default40 as SafeAreaView, default41 as ScrollView, default30 as SectionList, default22 as Share, default42 as StatusBar, default25 as Switch, default43 as Text, default44 as TextInput, default28 as TouchableHighlight, default29 as TouchableNativeFeedback, default31 as TouchableOpacity, default32 as TouchableWithoutFeedback, UIManager, default23 as Vibration, default45 as View, default26 as VirtualizedList, canUseDOM, dismissKeyboard, isWebColor, mergeRefs, processColor, default3 as render, requireNativeComponent, default2 as unstable_createElement, default47 as useColorScheme, useEvent, useHover, useLocaleContext, useMergeRefs, default48 as useWindowDimensions };
