// @tamagui/react-native-svg/dist/esm/index.native.js
import { createElement, forwardRef } from "react";
var Circle = forwardRef(function (p, ref) {
    return createElement("circle", {
      ...p,
      ref
    });
  }),
  ClipPath = forwardRef(function (p, ref) {
    return createElement("clipPath", {
      ...p,
      ref
    });
  }),
  Defs = forwardRef(function (p, ref) {
    return createElement("defs", {
      ...p,
      ref
    });
  }),
  Ellipse = forwardRef(function (p, ref) {
    return createElement("ellipse", {
      ...p,
      ref
    });
  }),
  ForeignObject = forwardRef(function (p, ref) {
    return createElement("foreignObject", {
      ...p,
      ref
    });
  }),
  G = forwardRef(function (p, ref) {
    return createElement("g", {
      ...p,
      ref
    });
  }),
  Image = forwardRef(function (p, ref) {
    return createElement("image", {
      ...p,
      ref
    });
  }),
  Line = forwardRef(function (p, ref) {
    return createElement("line", {
      ...p,
      ref
    });
  }),
  LinearGradient = forwardRef(function (p, ref) {
    return createElement("linearGradient", {
      ...p,
      ref
    });
  }),
  Marker = forwardRef(function (p, ref) {
    return createElement("marker", {
      ...p,
      ref
    });
  }),
  Mask = forwardRef(function (p, ref) {
    return createElement("mask", {
      ...p,
      ref
    });
  }),
  Path = forwardRef(function (p, ref) {
    return createElement("path", {
      ...p,
      ref
    });
  }),
  Pattern = forwardRef(function (p, ref) {
    return createElement("pattern", {
      ...p,
      ref
    });
  }),
  Polygon = forwardRef(function (p, ref) {
    return createElement("polygon", {
      ...p,
      ref
    });
  }),
  Polyline = forwardRef(function (p, ref) {
    return createElement("polyline", {
      ...p,
      ref
    });
  }),
  RadialGradient = forwardRef(function (p, ref) {
    return createElement("radialGradient", {
      ...p,
      ref
    });
  }),
  Rect = forwardRef(function (p, ref) {
    return createElement("rect", {
      ...p,
      ref
    });
  }),
  Shape = forwardRef(function (p, ref) {
    return createElement("shape", {
      ...p,
      ref
    });
  }),
  Stop = forwardRef(function (p, ref) {
    return createElement("stop", {
      ...p,
      ref
    });
  }),
  Svg = forwardRef(function (p, ref) {
    return createElement("svg", {
      ...p,
      ref
    });
  }),
  Symbol = forwardRef(function (p, ref) {
    return createElement("symbol", {
      ...p,
      ref
    });
  }),
  Text = forwardRef(function (p, ref) {
    return createElement("text", {
      ...p,
      ref
    });
  }),
  TextPath = forwardRef(function (p, ref) {
    return createElement("clipPath", {
      ...p,
      ref
    });
  }),
  TSpan = forwardRef(function (p, ref) {
    return createElement("tSpan", {
      ...p,
      ref
    });
  }),
  Use = forwardRef(function (p, ref) {
    return createElement("use", {
      ...p,
      ref
    });
  }),
  index_default = Svg;
export { Circle, ClipPath, Defs, Ellipse, ForeignObject, G, Image, Line, LinearGradient, Marker, Mask, Path, Pattern, Polygon, Polyline, RadialGradient, Rect, Shape, Stop, Svg, Symbol, TSpan, Text, TextPath, Use, index_default as default };
