// @tamagui/react-native-web-lite/dist/esm/vendor/react-native/vendor/emitter/EventEmitter.native.js
function _class_call_check(instance, Constructor) {
  if (!(instance instanceof Constructor)) throw new TypeError("Cannot call a class as a function");
}
function _defineProperties(target, props) {
  for (var i = 0; i < props.length; i++) {
    var descriptor = props[i];
    descriptor.enumerable = descriptor.enumerable || !1, descriptor.configurable = !0, "value" in descriptor && (descriptor.writable = !0), Object.defineProperty(target, descriptor.key, descriptor);
  }
}
function _create_class(Constructor, protoProps, staticProps) {
  return protoProps && _defineProperties(Constructor.prototype, protoProps), staticProps && _defineProperties(Constructor, staticProps), Constructor;
}
function _define_property(obj, key, value) {
  return key in obj ? Object.defineProperty(obj, key, {
    value,
    enumerable: !0,
    configurable: !0,
    writable: !0
  }) : obj[key] = value, obj;
}
var EventSubscription = /* @__PURE__ */function () {
    "use strict";

    function EventSubscription2() {
      _class_call_check(this, EventSubscription2);
    }
    return _create_class(EventSubscription2, [{
      key: "remove",
      value: function () {}
    }]), EventSubscription2;
  }(),
  IEventEmitter = /* @__PURE__ */function () {
    "use strict";

    function IEventEmitter2() {
      _class_call_check(this, IEventEmitter2);
    }
    return _create_class(IEventEmitter2, [{
      key: "addListener",
      value: function (eventType, listener, context) {
        return new EventSubscription();
      }
    }, {
      key: "emit",
      value: function (eventType) {
        for (var _len = arguments.length, args = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) args[_key - 1] = arguments[_key];
      }
    }, {
      key: "removeAllListeners",
      value: function (eventType) {}
    }, {
      key: "listenerCount",
      value: function (eventType) {
        return 0;
      }
    }]), IEventEmitter2;
  }();
var EventEmitter = /* @__PURE__ */function () {
  "use strict";

  function EventEmitter2() {
    _class_call_check(this, EventEmitter2), _define_property(this, "_registry", {});
  }
  return _create_class(EventEmitter2, [{
    /**
    * Registers a listener that is called when the supplied event is emitted.
    * Returns a subscription that has a `remove` method to undo registration.
    */
    key: "addListener",
    value: function (eventType, listener, context) {
      var registrations = allocate(this._registry, eventType),
        registration = {
          context,
          listener,
          remove() {
            registrations.delete(registration);
          }
        };
      return registrations.add(registration), registration;
    }
  }, {
    /**
    * Emits the supplied event. Additional arguments supplied to `emit` will be
    * passed through to each of the registered listeners.
    *
    * If a listener modifies the listeners registered for the same event, those
    * changes will not be reflected in the current invocation of `emit`.
    */
    key: "emit",
    value: function (eventType) {
      for (var _len = arguments.length, args = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) args[_key - 1] = arguments[_key];
      var registrations = this._registry[eventType];
      if (registrations != null) for (var _i = 0, _iter = [...registrations]; _i < _iter.length; _i++) {
        var registration = _iter[_i];
        registration.listener.apply(registration.context, args);
      }
    }
  }, {
    /**
    * Removes all registered listeners.
    */
    key: "removeAllListeners",
    value: function (eventType) {
      eventType == null ? this._registry = {} : delete this._registry[eventType];
    }
  }, {
    /**
    * Returns the number of registered listeners for the supplied event.
    */
    key: "listenerCount",
    value: function (eventType) {
      var registrations = this._registry[eventType];
      return registrations == null ? 0 : registrations.size;
    }
  }]), EventEmitter2;
}();
function allocate(registry, eventType) {
  var registrations = registry[eventType];
  return registrations == null && (registrations = /* @__PURE__ */new Set(), registry[eventType] = registrations), registrations;
}
export { EventSubscription, IEventEmitter, EventEmitter as default };
