// @tamagui/react-native-web-internals/dist/esm/modules/UIManager/index.native.js
import { measure, measureInWindow } from "@tamagui/use-element-layout";
var focusableElements = {
    A: !0,
    INPUT: !0,
    SELECT: !0,
    TEXTAREA: !0
  },
  UIManager = {
    blur(node) {
      try {
        node.blur();
      } catch {}
    },
    //
    focus(node) {
      try {
        var name = node.nodeName;
        node.getAttribute("tabIndex") == null && focusableElements[name] == null && node.setAttribute("tabIndex", "-1"), node.focus();
      } catch {}
    },
    //
    measure(node, callback) {
      return measure(node, callback);
    },
    measureInWindow(node, callback) {
      return measureInWindow(node, callback);
    },
    // note its flipped fail and success on purpose lol
    async measureLayout(node, relativeToNativeNode, onFail, onSuccess) {
      return measureLayout(node, relativeToNativeNode, onSuccess);
    },
    configureNextLayoutAnimation(config, onAnimationDidEnd) {
      onAnimationDidEnd();
    },
    // mocks
    setLayoutAnimationEnabledExperimental() {}
  },
  UIManager_default = UIManager;
export { UIManager_default as default };
