// @tamagui/react-native-web-lite/dist/cjs/vendor/react-native/Animated/Animated.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf, __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: !0 }) : target,
  mod
)), __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var Animated_exports = {};
__export(Animated_exports, {
  default: () => Animated_default
});
module.exports = __toCommonJS(Animated_exports);
var import_react_native_web_internals = require("@tamagui/react-native-web-internals"), import_AnimatedFlatList = __toESM(require("./components/AnimatedFlatList")), import_AnimatedImage = __toESM(require("./components/AnimatedImage")), import_AnimatedScrollView = __toESM(require("./components/AnimatedScrollView")), import_AnimatedSectionList = __toESM(require("./components/AnimatedSectionList")), import_AnimatedText = __toESM(require("./components/AnimatedText")), import_AnimatedView = __toESM(require("./components/AnimatedView")), import_AnimatedMock = __toESM(require("./AnimatedMock")), import_AnimatedImplementation = __toESM(require("./AnimatedImplementation"));
const Animated = import_react_native_web_internals.Platform.isTesting ? import_AnimatedMock.default : import_AnimatedImplementation.default;
var Animated_default = {
  FlatList: import_AnimatedFlatList.default,
  Image: import_AnimatedImage.default,
  ScrollView: import_AnimatedScrollView.default,
  SectionList: import_AnimatedSectionList.default,
  Text: import_AnimatedText.default,
  View: import_AnimatedView.default,
  ...Animated
};
