// @tamagui/react-native-web-internals/dist/cjs/modules/InteractionManager.native.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf, __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: !0 }) : target,
  mod
)), __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var InteractionManager_exports = {};
__export(InteractionManager_exports, {
  InteractionManager: () => InteractionManager
});
module.exports = __toCommonJS(InteractionManager_exports);
var import_invariant = require("./invariant"), import_requestIdleCallback = __toESM(require("./requestIdleCallback/index"));
function _class_call_check(instance, Constructor) {
  if (!(instance instanceof Constructor))
    throw new TypeError("Cannot call a class as a function");
}
function _defineProperties(target, props) {
  for (var i = 0; i < props.length; i++) {
    var descriptor = props[i];
    descriptor.enumerable = descriptor.enumerable || !1, descriptor.configurable = !0, "value" in descriptor && (descriptor.writable = !0), Object.defineProperty(target, descriptor.key, descriptor);
  }
}
function _create_class(Constructor, protoProps, staticProps) {
  return protoProps && _defineProperties(Constructor.prototype, protoProps), staticProps && _defineProperties(Constructor, staticProps), Constructor;
}
function _define_property(obj, key, value) {
  return key in obj ? Object.defineProperty(obj, key, {
    value,
    enumerable: !0,
    configurable: !0,
    writable: !0
  }) : obj[key] = value, obj;
}
var EventEmitter = /* @__PURE__ */ function() {
  "use strict";
  function EventEmitter2() {
    _class_call_check(this, EventEmitter2), _define_property(this, "_registry", {});
  }
  return _create_class(EventEmitter2, [
    {
      key: "addListener",
      value: function(eventType, listener, context) {
        var registrations = this._allocate(eventType), registration = {
          context,
          listener,
          remove: function() {
            registrations.delete(registration);
          }
        };
        return registrations.add(registration), registration;
      }
    },
    {
      key: "emit",
      value: function(eventType) {
        for (var _len = arguments.length, args = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++)
          args[_key - 1] = arguments[_key];
        var registrations = this._registry[eventType];
        if (registrations != null) {
          var _iteratorNormalCompletion = !0, _didIteratorError = !1, _iteratorError = void 0;
          try {
            for (var _iterator = Array.from(registrations)[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = !0) {
              var registration = _step.value;
              registration.listener.apply(registration.context, args);
            }
          } catch (err) {
            _didIteratorError = !0, _iteratorError = err;
          } finally {
            try {
              !_iteratorNormalCompletion && _iterator.return != null && _iterator.return();
            } finally {
              if (_didIteratorError)
                throw _iteratorError;
            }
          }
        }
      }
    },
    {
      key: "_allocate",
      value: function(eventType) {
        var registrations = this._registry[eventType];
        return registrations == null && (registrations = /* @__PURE__ */ new Set(), this._registry[eventType] = registrations), registrations;
      }
    }
  ]), EventEmitter2;
}(), TaskQueue = /* @__PURE__ */ function() {
  "use strict";
  function TaskQueue2(param) {
    var { onMoreTasks } = param;
    _class_call_check(this, TaskQueue2), _define_property(this, "_queueStack", void 0), _define_property(this, "_onMoreTasks", void 0), this._onMoreTasks = onMoreTasks, this._queueStack = [
      {
        tasks: [],
        popable: !0
      }
    ];
  }
  return _create_class(TaskQueue2, [
    {
      key: "enqueueTasks",
      value: function(tasks) {
        var _this = this;
        tasks.forEach(function(task) {
          return _this._enqueue(task);
        });
      }
    },
    {
      key: "cancelTasks",
      value: function(tasksToCancel) {
        this._queueStack = this._queueStack.map(function(queue) {
          return {
            ...queue,
            tasks: queue.tasks.filter(function(task) {
              return !tasksToCancel.includes(task);
            })
          };
        }).filter(function(queue, idx) {
          return queue.tasks.length > 0 || idx === 0;
        });
      }
    },
    {
      key: "hasTasksToProcess",
      value: function() {
        return this._getCurrentQueue().length > 0;
      }
    },
    {
      key: "processNext",
      value: function() {
        var queue = this._getCurrentQueue();
        if (queue.length) {
          var task = queue.shift();
          try {
            typeof task == "object" && task && "gen" in task ? this._genPromise(task) : typeof task == "object" && task && "run" in task ? task.run() : ((0, import_invariant.invariant)(typeof task == "function", `Expected Function, SimpleTask, or PromiseTask, but got:
` + JSON.stringify(task, null, 2)), task());
          } catch (e) {
            if (e instanceof Error) {
              var taskName = task && typeof task == "object" && "name" in task ? task.name : "";
              e.message = "TaskQueue: Error with task " + taskName + ": " + e.message;
            }
            throw e;
          }
        }
      }
    },
    {
      key: "_enqueue",
      value: function(task) {
        this._getCurrentQueue().push(task);
      }
    },
    {
      key: "_getCurrentQueue",
      value: function() {
        var stackIdx = this._queueStack.length - 1, queue = this._queueStack[stackIdx];
        return queue.popable && queue.tasks.length === 0 && stackIdx > 0 ? (this._queueStack.pop(), this._getCurrentQueue()) : queue.tasks;
      }
    },
    {
      key: "_genPromise",
      value: function(task) {
        var _this = this, length = this._queueStack.push({
          tasks: [],
          popable: !1
        }), stackIdx = length - 1, stackItem = this._queueStack[stackIdx];
        task.gen().then(function() {
          stackItem.popable = !0, _this.hasTasksToProcess() && _this._onMoreTasks();
        }).catch(function(ex) {
          setTimeout(function() {
            throw ex instanceof Error && (ex.message = `TaskQueue: Error resolving Promise in task ${task.name}: ${ex.message}`), ex;
          }, 0);
        });
      }
    }
  ]), TaskQueue2;
}(), _emitter = new EventEmitter(), InteractionManager = {
  Events: {
    interactionStart: "interactionStart",
    interactionComplete: "interactionComplete"
  },
  /**
  * Schedule a function to run after all interactions have completed.
  */
  runAfterInteractions(task) {
    var tasks = [], promise = new Promise(function(resolve) {
      _scheduleUpdate(), task && tasks.push(task), tasks.push({
        run: resolve,
        name: "resolve " + (task && typeof task == "object" && "name" in task && task.name || "?")
      }), _taskQueue.enqueueTasks(tasks);
    });
    return {
      then: promise.then.bind(promise),
      done: promise.then.bind(promise),
      cancel: function() {
        _taskQueue.cancelTasks(tasks);
      }
    };
  },
  /**
  * Notify manager that an interaction has started.
  */
  createInteractionHandle() {
    _scheduleUpdate();
    var handle = ++_inc;
    return _addInteractionSet.add(handle), handle;
  },
  /**
  * Notify manager that an interaction has completed.
  */
  clearInteractionHandle(handle) {
    (0, import_invariant.invariant)(!!handle, "Must provide a handle to clear."), _scheduleUpdate(), _addInteractionSet.delete(handle), _deleteInteractionSet.add(handle);
  },
  addListener: _emitter.addListener.bind(_emitter),
  /**
  * Set deadline for task processing
  */
  setDeadline(deadline) {
    _deadline = deadline;
  }
}, _interactionSet = /* @__PURE__ */ new Set(), _addInteractionSet = /* @__PURE__ */ new Set(), _deleteInteractionSet = /* @__PURE__ */ new Set(), _taskQueue = new TaskQueue({
  onMoreTasks: _scheduleUpdate
}), _nextUpdateHandle = null, _inc = 0, _deadline = -1;
function _scheduleUpdate() {
  _nextUpdateHandle || (_deadline > 0 ? _nextUpdateHandle = setTimeout(_processUpdate) : _nextUpdateHandle = (0, import_requestIdleCallback.default)(_processUpdate));
}
function _processUpdate() {
  _nextUpdateHandle = null;
  var interactionCount = _interactionSet.size;
  _addInteractionSet.forEach(function(handle) {
    return _interactionSet.add(handle);
  }), _deleteInteractionSet.forEach(function(handle) {
    return _interactionSet.delete(handle);
  });
  var nextInteractionCount = _interactionSet.size;
  if (interactionCount !== 0 && nextInteractionCount === 0 ? _emitter.emit("interactionComplete") : interactionCount === 0 && nextInteractionCount !== 0 && _emitter.emit("interactionStart"), nextInteractionCount === 0) {
    for (var begin = Date.now(); _taskQueue.hasTasksToProcess(); )
      if (_taskQueue.processNext(), _deadline > 0 && Date.now() - begin >= _deadline) {
        _scheduleUpdate();
        break;
      }
  }
  _addInteractionSet.clear(), _deleteInteractionSet.clear();
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  InteractionManager
});
